src/main/java/net/spy/memcached/protocol/BaseOperationImpl.java

```diff
@@ -166,6 +169,19 @@ protected final synchronized void transitionState(OperationState newState) {
     if (state == OperationState.COMPLETE) {
       callback.complete();
     }
+
+    notifyStateObservers(prevState, state);
+  }
+
+  private void notifyStateObservers(OperationState prevState, OperationState newState) {
+    for (OperationStateChangeObserver observer : stateObservers) {
+      observer.stateChanged(prevState, newState);
+    }
+  }
+
+  public final void addStateObserver(OperationStateChangeObserver observer) {
```

Comment:
Should synchronize?

Reply:
currently no, do you want add anyway?

Reply:
Not sure, when you're going to call addStateObserver?

Reply:
in the base operations factory, when operation is created, before it is returned to the caller